tensorflow/core/kernels/sparse_tensor_dense_matmul_op_gpu.cu.cc

```diff
@@ -71,62 +101,111 @@ __global__ void SparseTensorDenseMatMulKernel(
 
-    // TODO(ebrevdo): Should this be alpha * nnz instead of
-    // out.size()?  Perhaps p * nnz ?
-    GpuLaunchConfig config = GetGpuLaunchConfig(p * nnz, d);
-
-    TF_CHECK_OK(GpuLaunchKernel(
-        SparseTensorDenseMatMulKernel<T, Tsum, Tindices, ADJ_A, ADJ_B>,
-        config.block_count, config.thread_per_block, 0, d.stream(), nnz, m,
-        b_rows, b_cols, p, a_indices.data(), a_values.data(), b.data(),
-        maybe_temp_out_data));
-
-    if (sum_type_is_different) {
-      out.device(d) = temp_out_t.matrix<Tsum>().template cast<T>();
+    if (RequireDeterminism()) {
+      if (std::is_same<T, double>::value ||
+          std::is_same<T, complex128>::value) {
+        return errors::Unimplemented(
+            "No deterministic GPU implementation of sparse_dense_matmul "
+            "available for data of type tf.float64 or tf.complex128");
+      }
+      using Tsum = typename SumType<T>::type_for_determinism;
+      ComputeWithPossiblyHigherPrecisionAccumulator <Tsum>(ctx, out, a_indices,
+                                                 a_values, b);
+    } else {
+      using Tsum = typename SumType<T>::type;
+      ComputeWithPossiblyHigherPrecisionAccumulator <Tsum>(ctx, out, a_indices,
+                                                 a_values, b);
     }
-
     return Status::OK();
   }
+
+  private:
+    template <typename Tsum>
+    static EIGEN_ALWAYS_INLINE Status
+    ComputeWithPossiblyHigherPrecisionAccumulator (
+        OpKernelContext* ctx, typename TTypes<T>::Matrix &out,
+        typename TTypes<Tindices>::ConstMatrix a_indices,
+        typename TTypes<T>::ConstVec a_values,
+        typename TTypes<T>::ConstMatrix b) {
+
+      int nnz = a_values.size();
+      // out = A * B, A is [m x n] and B is [n x p], out is [m x p]
+      int m = out.dimension(0);
+      int p = out.dimension(1);
+      int b_rows = b.dimension(0);
+      int b_cols = b.dimension(1);
+      const GPUDevice& d = ctx->eigen_device<GPUDevice>();
+      GpuLaunchConfig config = GetGpuLaunchConfig(p * nnz, d);
+
+      Tensor temp_out_t;
+      Tsum* maybe_temp_out_data = nullptr;
+      bool sum_type_is_different = !std::is_same<T, Tsum>::value;
+
+      if (sum_type_is_different) {
+        TF_RETURN_IF_ERROR(ctx->allocate_temp(
+            DataTypeToEnum<Tsum>::value,
+            TensorShape({out.dimension(0), out.dimension(1)}), &temp_out_t));
+
+        auto temp_out = temp_out_t.matrix<Tsum>();
+        maybe_temp_out_data = temp_out.data();
+      } else {
+        // Note: The reinterpret cast is only required to avoid a compilation
+        // error; it is only used if Tsum == T.
+        maybe_temp_out_data = reinterpret_cast<Tsum*>(out.data());
+      }
+
+      TF_CHECK_OK(GpuLaunchKernel(
+          SetZero<Tsum>, config.block_count, config.thread_per_block, 0,
+          d.stream(), m * p, maybe_temp_out_data));
+
+      // TODO(ebrevdo): Should this be alpha * nnz instead of
+      // out.size()?  Perhaps p * nnz ?
```

Comment:
Do you understand what this comment means or what it refers to? It looks like it might be a zombie.

Reply:
I don't understand. Will delete it.

Reply:
I found the comment in the original code:

```
    // TODO(ebrevdo): Should this be alpha * nnz instead of
    // out.size()?  Perhaps p * nnz ?
    GpuLaunchConfig config = GetGpuLaunchConfig(p * nnz, d);
```

The comment makes sense with the line of code it refers to. It looks like the TODO was completed (perhaps by someone other than @ebrevdo).